pkg/sync2/set.go

```diff
@@ -0,0 +1,223 @@
+package sync2
+
+import (
+	"fmt"
+	"strings"
+
+	"gopkg.in/typ.v3/pkg/sets"
+)
+
+// NewSetFromSlice returns a Set with all values from a slice added to it.
+func NewSetFromSlice[S ~[]E, E comparable](slice S) sets.Set[E] {
+	var set Set[E]
+	for _, v := range slice {
+		set.Add(v)
+	}
+	return &set
+}
+
+// NewSetFromKeys returns a Set with all keys from a map added to it.
+func NewSetFromKeys[M ~map[K]V, K comparable, V any](m M) sets.Set[K] {
+	var set Set[K]
+	for k := range m {
+		set.Add(k)
+	}
+	return &set
+}
+
+// NewSetFromValues returns a Set with all values from a map added to it.
+func NewSetFromValues[M ~map[K]V, K comparable, V comparable](m M) sets.Set[V] {
+	var set Set[V]
+	for _, v := range m {
+		set.Add(v)
+	}
+	return &set
+}
+
+// Set holds a collection of values with no duplicates. Its methods are based
+// on the mathematical branch of set theory, and its implementation is using a
+// Map[T, struct{}].
+type Set[T comparable] struct {
+	m Map[T, struct{}]
+}
+
+// assert that Set implements sets.Set interface.
+var _ sets.Set[int] = &Set[int]{}
+
+// String converts this set to its string representation.
+func (s *Set[T]) String() string {
+	var sb strings.Builder
+	sb.WriteByte('{')
+	addDelim := false
+	s.Range(func(value T) bool {
+		if addDelim {
+			sb.WriteByte(' ')
+		} else {
+			addDelim = true
+		}
+		fmt.Fprint(&sb, value)
+		return true
+	})
+	sb.WriteByte('}')
+	return sb.String()
+}
+
+// Has returns true if the value exists in the set.
+func (s *Set[T]) Has(value T) bool {
+	_, has := s.m.Load(value)
+	return has
+}
+
+// Add will add an element to the set, and return true if it was added
+// or false if the value already existed in the set.
+func (s *Set[T]) Add(value T) bool {
+	if s.Has(value) {
+		return false
+	}
+	s.m.Store(value, struct{}{})
+	return true
+}
+
+// AddSet will add all element found in specified set to this set, and
+// return the number of values that was added.
+func (s *Set[T]) AddSet(set sets.Set[T]) int {
+	var added int
+	set.Range(func(value T) bool {
+		if s.Add(value) {
+			added++
+		}
+		return true
+	})
+	return added
+}
+
+// Remove will remove an element from the set, and return true if it was removed
+// or false if no such value existed in the set.
+func (s *Set[T]) Remove(value T) bool {
+	if !s.Has(value) {
+		return false
+	}
+	s.m.Delete(value)
+	return true
+}
+
+// RemoveSet will remove all element found in specified set from this set, and
+// return the number of values that was removed.
+func (s *Set[T]) RemoveSet(set sets.Set[T]) int {
+	var removed int
+	set.Range(func(value T) bool {
+		if s.Remove(value) {
+			removed++
+		}
+		return true
+	})
+	return removed
+}
+
+// Clone returns a copy of the set.
+func (s *Set[T]) Clone() sets.Set[T] {
+	var clone Set[T]
+	clone.AddSet(s)
+	return &clone
+}
+
+// Slice returns a new slice of all values in the set.
+func (s *Set[T]) Slice() []T {
+	result := make([]T, 0)
```

Comment:
No need to initialize the slice if you're setting it to zero anyway

```suggestion
	var result []T
```